### PR TITLE
[Xamarin.Android.Build.Tasks] Validate the `android:versionCode`

### LINKED ISSUE #2023: Input string was not in correct format
File: Xamarin.Android.Common.targets
Line: 2007

### File Log:

```
/Library/Frameworks/Mono.framework/External/xbuild/Xamarin/Android/Xamarin.Android.Common.targets(3,3): Error: Input string was not in a correct format.
  at System.Number.StringToNumber (System.String str, System.Globalization.NumberStyles options, System.Number+NumberBuffer& number, System.Globalization.NumberFormatInfo info, System.Boolean parseDecimal) [0x0005f] in /Users/builder/jenkins/workspace/build-package-osx-mono/2017-12/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/number.cs:1077 
  at System.Number.ParseInt32 (System.String s, System.Globalization.NumberStyles style, System.Globalization.NumberFormatInfo info) [0x00013] in /Users/builder/jenkins/workspace/build-package-osx-mono/2017-12/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/number.cs:745 
  at System.Int32.Parse (System.String s) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2017-12/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/int32.cs:120 
  at Xamarin.Android.Tasks.ManifestDocument.CalculateVersionCode (System.String currentAbi, System.String versionCodePattern, System.String versionCodeProperties) [0x000bb] in <96027da06dd5411688a31011072e256f>:0 
  at Xamarin.Android.Tasks.Aapt.GenerateCommandLineCommands (System.String ManifestFile, System.String currentAbi, System.String currentResourceOutputFile) [0x000d7] in <96027da06dd5411688a31011072e256f>:0 
  at Xamarin.Android.Tasks.Aapt.ProcessManifest (Microsoft.Build.Framework.ITaskItem manifestFile) [0x0017d] in <96027da06dd5411688a31011072e256f>:0 
  at System.Threading.Tasks.Parallel+<>c__DisplayClass30_0`2[TSource,TLocal].<ForEachWorker>b__0 (System.Int32 i) [0x00000] in <bb7b695b8c6246b3ac1646577aea7650>:0 
  at System.Threading.Tasks.Parallel+<>c__DisplayClass17_0`1[TLocal].<ForWorker>b__1 () [0x0020b] in <bb7b695b8c6246b3ac1646577aea7650>:0 
  at System.Threading.Tasks.Task.InnerInvoke () [0x0000f] in /Users/builder/jenkins/workspace/build-package-osx-mono/2017-12/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/Tasks/Task.cs:2886 
  at System.Threading.Tasks.Task.InnerInvokeWithArg (System.Threading.Tasks.Task childTask) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2017-12/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/Tasks/Task.cs:2910 
  at System.Threading.Tasks.Task+<>c__DisplayClass178_0.<ExecuteSelfReplicating>b__0 (System.Object <p0>) [0x00086] in /Users/builder/jenkins/workspace/build-package-osx-mono/2017-12/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/Tasks/Task.cs:2635  (DistributorRoute.Android)
```

### PR DESCRIPTION
**Pull request by dellis1972**
Fixes #2023

The `versionCode` in the android manifest should be
an `int` value. It should fall within the range of
`0` to `2100000000`.

This commit adds some checks which will emit MSBuild errors
if we encounter an invalid `versionCode`. It also
adds some unit tests to test for decimal values and
the boundary conditions.

### CODE REVIEW

File: src/Xamarin.Android.Build.Tasks/Tasks/Aapt.cs [hunk 1]
```diff
@@ -201,7 +201,10 @@ void ProcessManifest (ITaskItem manifestFile)
 				var currentResourceOutputFile = abi != null ? string.Format ("{0}-{1}", ResourceOutputFile, abi) : ResourceOutputFile;
 				if (!string.IsNullOrEmpty (currentResourceOutputFile) && !Path.IsPathRooted (currentResourceOutputFile))
 					currentResourceOutputFile = Path.Combine (WorkingDirectory, currentResourceOutputFile);
-				if (!ExecuteForAbi (GenerateCommandLineCommands (manifest, abi, currentResourceOutputFile), currentResourceOutputFile)) {
+				string cmd = GenerateCommandLineCommands (manifest, abi, currentResourceOutputFile);
+				if (string.IsNullOrWhiteSpace (cmd))
+					Cancel ();
+				if (!ExecuteForAbi (cmd, currentResourceOutputFile)) {
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Tasks/Aapt.cs]: maybe this would be cleaner as:
```csharp
if (string.IsNullOrWhiteSpace (cmd) || !ExecuteForAbi (cmd, currentResourceOutputFile)) {
```

File: src/Xamarin.Android.Build.Tasks/Tasks/Aapt2Link.cs [hunk 2]
```diff
@@ -271,7 +275,10 @@ void ProcessManifest (ITaskItem manifestFile)
 				var currentResourceOutputFile = abi != null ? string.Format ("{0}-{1}", outputFile, abi) : outputFile;
 				if (!string.IsNullOrEmpty (currentResourceOutputFile) && !Path.IsPathRooted (currentResourceOutputFile))
 					currentResourceOutputFile = Path.Combine (WorkingDirectory, currentResourceOutputFile);
-				if (!ExecuteForAbi (GenerateCommandLineCommands (manifest, abi, currentResourceOutputFile), currentResourceOutputFile)) {
+				string cmd = GenerateCommandLineCommands (manifest, abi, currentResourceOutputFile);
+				if (string.IsNullOrWhiteSpace (cmd))
+					Cancel ();
+				if (!ExecuteForAbi (cmd, currentResourceOutputFile)) {
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Tasks/Aapt2Link.cs]: same here:
```csharp
if (string.IsNullOrWhiteSpace (cmd) || !ExecuteForAbi (cmd, currentResourceOutputFile)) {
```

File: src/Xamarin.Android.Build.Tasks/Tasks/Aapt.cs [hunk 3]
```diff
@@ -287,6 +290,10 @@ protected string GenerateCommandLineCommands (string ManifestFile, string curren
 			} else if (!string.IsNullOrEmpty (VersionCodePattern)) {
 				manifest.CalculateVersionCode (null, VersionCodePattern, VersionCodeProperties);
 			}
+			if (!manifest.ValidateVersionCode (out string error)) {
+				Log.LogCodedError ("APT0000", ManifestFile, 0, error);
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tasks/Aapt.cs]: An invalid `versionCode` value should not be an `APT0000` error, so that we can distinguish these failures in telemetry.  We should introduce a new XA.... code, presumably XA0002? https://github.com/xamarin/xamarin-android/tree/master/Documentation/guides/messages

File: Documentation/guides/messages/README.md [hunk 4]
```diff
@@ -19,7 +19,9 @@
 
 + [XA0000](xa0000.md): Could not determine $(AndroidApiLevel) or $(TargetFrameworkVersion).
 + [XA0001](xa0001.md): Invalid or unsupported `$(TargetFrameworkVersion)` value.
-+ [XA0118](xa0002.md): Could not find mono.android.jar
++ [XA0002](xa0002.md): Could not find mono.android.jar
```
Comment (dellis1972) [inline Documentation/guides/messages/README.md]: not sure why we had the wrong code here... Also XA0002 does not appear to be used anywhere in the repo...
